File: utils/security.js
const bcrypt = require('bcrypt');
const jwt = require('jsonwebtoken');
const crypto = require('crypto');
const dotenv = require('dotenv');

// Load environment variables.  If no `.env` file exists the
// defaults from `.env` will be used when executing via
// npm scripts because dotenv doesn't set any values for undefined
// variables; instead we rely on destructuring defaults below.
dotenv.config();

const {
  JWT_SECRET = 'change-me',
  ACCESS_MIN = '15',
  REFRESH_DAYS = '15'
} = process.env;

/**
 * Hash a plain text password using bcrypt.  A salt is generated
 * automatically.  Returns a promise that resolves to the hashed
 * password.
 *
 * @param {string} plain The raw password
 * @returns {Promise<string>} The bcrypt hash
 */
async function hashPassword(plain) {
  const saltRounds = 12;
  return bcrypt.hash(plain, saltRounds);
}

/**
 * Verify a password against a bcrypt hash.
 *
 * @param {string} plain The raw password
 * @param {string} hash The stored bcrypt hash
 * @returns {Promise<boolean>} True if the password matches
 */
async function verifyPassword(plain, hash) {
  return bcrypt.compare(plain, hash);
}

/**
 * Create a signed JWT access token.  The payload contains the
 * subject (user id) and roles.  The token expires after
 * ``ACCESS_MIN`` minutes.
 *
 * @param {string} sub User identifier
 * @param {string[]} roles List of role names
 * @returns {string} A signed JWT
 */
function createAccessToken(sub, roles = []) {
  const exp = Math.floor(Date.now() / 1000) + parseInt(ACCESS_MIN, 10) * 60;
  const payload = { sub, roles, exp };
  return jwt.sign(payload, JWT_SECRET, { algorithm: 'HS256' });
}

/**
 * Decode and verify a JWT access token.  Throws if invalid or
 * expired.  Returns the payload object if valid.
 *
 * @param {string} token The JWT to decode
 * @returns {object} The decoded payload
 */
function decodeAccessToken(token) {
  return jwt.verify(token, JWT_SECRET, { algorithms: ['HS256'] });
}

/**
 * Generate a refresh token and its digest.  The raw token should
 * be returned to the client while the digest is stored in the
 * database.  We use Node's crypto API to generate a high
 * entropy token and compute a SHA256 hex digest.
 *
 * @returns {{ raw: string, digest: string }} The token and digest
 */
function makeRefreshToken() {
  const raw = crypto.randomBytes(48).toString('base64url');
  const digest = crypto.createHash('sha256').update(raw).digest('hex');
  return { raw, digest };
}

/**
 * Compute the expiry date for a refresh token based on
 * ``REFRESH_DAYS``.  Returns a JavaScript Date object set to the
 * expiration time in UTC.
 *
 * @returns {Date} The expiry timestamp
 */
function refreshExp() {
  const days = parseInt(REFRESH_DAYS, 10) || 15;
  const expires = new Date();
  expires.setUTCDate(expires.getUTCDate() + days);
  return expires;
}

module.exports = {
  hashPassword,
  verifyPassword,
  createAccessToken,
  decodeAccessToken,
  makeRefreshToken,
  refreshExp
};

File: utils/security.py
# utils/security.py
from __future__ import annotations

from datetime import datetime, timedelta, timezone
from hashlib import sha256
from secrets import token_urlsafe
from typing import List, Optional

import jwt
from passlib.context import CryptContext

from config import settings

# ---------------------------
# Password hashing
# ---------------------------
# Use bcrypt_sha256 to avoid bcrypt's 72-byte input limit.
# Keep "bcrypt" to verify legacy hashes stored earlier.
pwd_ctx = CryptContext(
    schemes=["bcrypt_sha256", "bcrypt"],
    deprecated="auto",
    bcrypt__default_rounds=12,
)


def hash_password(raw: str) -> str:
    """Hash password using bcrypt_sha256 (preferred)."""
    return pwd_ctx.hash(raw)


def verify_password(raw: str, hashed: str) -> bool:
    """Verify password against a stored hash (supports bcrypt_sha256 & legacy bcrypt)."""
    return pwd_ctx.verify(raw, hashed)


def needs_rehash(hashed: str) -> bool:
    """Should the stored hash be upgraded (e.g., legacy bcrypt -> bcrypt_sha256)?"""
    return pwd_ctx.needs_update(hashed)


def maybe_rehash_after_verify(raw: str, hashed: str) -> Optional[str]:
    """
    If verify succeeds and the stored hash is legacy/weaker, return an upgraded hash
    you can persist. Otherwise return None.
    """
    if verify_password(raw, hashed) and needs_rehash(hashed):
        return hash_password(raw)
    return None


# ---------------------------
# JWT helpers
# ---------------------------
ALGORITHM = "HS256"


def create_access_token(
        sub: str,
        roles: Optional[List[str]] = None,
        minutes: Optional[int] = None,
) -> str:
    """
    Create a short-lived access token.
      - sub: subject (user id as string)
      - roles: list of role strings
      - minutes: override lifetime (defaults to settings.ACCESS_MIN)
    """
    exp_minutes = minutes if minutes is not None else settings.ACCESS_MIN
    now = datetime.now(timezone.utc)
    payload = {
        "sub": sub,
        "roles": roles or [],
        "iat": int(now.timestamp()),
        "exp": int((now + timedelta(minutes=exp_minutes)).timestamp()),
    }
    return jwt.encode(payload, settings.JWT_SECRET, algorithm=ALGORITHM)


def decode_access_token(token: str) -> dict:
    """Decode & validate an access token (raises if invalid/expired)."""
    return jwt.decode(token, settings.JWT_SECRET, algorithms=[ALGORITHM])


# ---------------------------
# Refresh token helpers
# ---------------------------
def make_refresh_token() -> dict:
    """
    Generate a refresh token pair:
      {"raw": <opaque-token-for-client>, "digest": <sha256-for-db>}
    """
    raw = token_urlsafe(48)
    digest = sha256(raw.encode("utf-8")).hexdigest()
    return {"raw": raw, "digest": digest}


def refresh_exp(days: Optional[int] = None) -> datetime:
    """Compute refresh token expiry (UTC)."""
    d = days if days is not None else settings.REFRESH_DAYS
    return datetime.now(timezone.utc) + timedelta(days=d)


# ---------------------------
# Role normalisation
# ---------------------------
def normalise_role(v: Optional[str]) -> Optional[str]:
    """
    Canonicalise role:
      - trim
      - spaces/underscores -> hyphens
      - uppercase
      - SUPERADMIN / SUPER_ADMIN -> SUPER-ADMIN
    """
    if not v:
        return None
    r = v.strip().replace(" ", "-").replace("_", "-").upper()
    if r in ("SUPERADMIN", "SUPER-ADMIN", "SUPER_ADMIN"):
        r = "SUPER-ADMIN"
    return r
